runners/flink/runner/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java

```diff
@@ -997,6 +1124,81 @@ public void flatMap(
     }
   }
 
+  private static class GBKIntoKeyedWorkItemsTranslator<K, InputT>
+      extends FlinkStreamingPipelineTranslator.StreamTransformTranslator<
+      SplittableParDo.GBKIntoKeyedWorkItems<K, InputT>> {
+
+    @Override
+    boolean canTranslate(
+        SplittableParDo.GBKIntoKeyedWorkItems<K, InputT> transform,
+        FlinkStreamingTranslationContext context) {
+      return true;
+    }
+
+    @Override
+    public void translateNode(
+        SplittableParDo.GBKIntoKeyedWorkItems<K, InputT> transform,
+        FlinkStreamingTranslationContext context) {
+
+      PCollection<KV<K, InputT>> input = context.getInput(transform);
+
+      KvCoder<K, InputT> inputKvCoder = (KvCoder<K, InputT>) input.getCoder();
+
+      SingletonKeyedWorkItemCoder<K, InputT> workItemCoder = SingletonKeyedWorkItemCoder.of(
+          inputKvCoder.getKeyCoder(),
+          inputKvCoder.getValueCoder(),
+          input.getWindowingStrategy().getWindowFn().windowCoder());
+
+      DataStream<WindowedValue<KV<K, InputT>>> inputDataStream = context.getInputDataStream(input);
+
+      WindowedValue.
+          FullWindowedValueCoder<SingletonKeyedWorkItem<K, InputT>> windowedWorkItemCoder =
+          WindowedValue.getFullCoder(
+              workItemCoder,
+              input.getWindowingStrategy().getWindowFn().windowCoder());
+
+      CoderTypeInformation<WindowedValue<SingletonKeyedWorkItem<K, InputT>>> workItemTypeInfo =
+          new CoderTypeInformation<>(windowedWorkItemCoder);
+
+      DataStream<WindowedValue<SingletonKeyedWorkItem<K, InputT>>> workItemStream =
+          inputDataStream
+              .flatMap(new ToKeyedWorkItem<K, InputT>())
+              .returns(workItemTypeInfo).name("ToKeyedWorkItem");
+
+      KeyedStream<
+          WindowedValue<
+              SingletonKeyedWorkItem<K, InputT>>, ByteBuffer> keyedWorkItemStream = workItemStream
+          .keyBy(new WorkItemKeySelector<K, InputT>(inputKvCoder.getKeyCoder()));
+
+        context.setOutputDataStream(context.getOutput(transform), keyedWorkItemStream);
+    }
+
+    private static class ToKeyedWorkItem<K, InputT>
+        extends RichFlatMapFunction<
+        WindowedValue<KV<K, InputT>>,
+        WindowedValue<SingletonKeyedWorkItem<K, InputT>>> {
+
+      @Override
+      public void flatMap(
+          WindowedValue<KV<K, InputT>> inWithMultipleWindows,
+          Collector<WindowedValue<SingletonKeyedWorkItem<K, InputT>>> out) throws Exception {
+
+        // we need to wrap each one work item per window for now
+        // since otherwise the PushbackSideInputRunner will not correctly
```

Comment:
The reason behind this is the interplay between `GroupAlsoByWindowViaWindowSetDoFn`/`GroupAlsoByWindowViaWindowSetNewDoFn` and `PushbackSideInputDoFnRunner`.

I'm afraid we might need some refactoring/fixing there, now that I'm looking at it: 

1. `GroupAlsoByWindowViaWindowSetDoFn` is a `DoFn<KeyedWorkItem<K, InputT>, KV<K, OutputT>>` so you have to push in `KeyedWorkItems`. These themselves contain `WindowedValue<InputT>` (or timers).
2. For executing a `DoFn` we use a `DoFnRunner`. For our problem the interesting case is using a `PushbackSideInputDoFnRunner`. The interesting method is `processElementInReadyWindows(WindowedValue<InputT> elem)` where `InputT` is the input type of the `DoFn` which, for the windowing case, is `KeyedWorkItem<K, InputT>` (from above). The actual expanded type signature is thus `processElementInReadyWindows(WindowedValue<KeyedWorkItem<K, InputT>> elem)` where the keyed work items again contain `WindowedValues` (again, from above). 

I think the `PushbackSideInputDoFnRunner` was not initially meant for executing `GroupAlsoByWindowViaWindowSetDoFns`.

(That's the explanation for why we have this code for the `GroupByKey` translation, for translating a stateful/splittable `DoFn` it's not necessary, IMO.